letsencrypt/client/dns_authenticator.py

```diff
@@ -0,0 +1,261 @@
+"""DNS Authenticator"""
+import dns.query
+import dns.tsigkeyring
+import dns.update
+import dns.resolver
+
+import zope.interface
+
+from letsencrypt.acme import challenges
+
+from letsencrypt.client import achallenges
+from letsencrypt.client import constants
+from letsencrypt.client import errors
+from letsencrypt.client import interfaces
+
+def find_valid_key(tsig_keys, domain):
+    """Search provided TSIG key pairs.
+
+    Search the keypairs in `config.dns_tsig_keys` for one valid for the
+    provided `domain`.
+
+    :param list tsig_keys: List of tuples containing key name, key secret,
+        and domains key is valid for.
+    :param str domain: Domain name to look for a key pair for.
+
+    :returns: Keyring valid for `domain` or `None`.
+    :rtype: dns.tsigkeyring
+
+    """
+    for keypair in tsig_keys:
+        if domain in keypair[2]:
+            return dns.tsigkeyring.from_text(
+                {keypair[0]: keypair[1]})
+
+def add_record(zone, token, keyring):
+    """Add record DNS request generator.
+
+    Create a `dns.message.Message` object asking the server to
+    create a TXT record for the challenge subdomain
+    (`_acme-challenge.example.com`) signed by the provided
+    TSIG keyring.
+
+    :param str zone: Zone (domain) in which the record should
+        be provisioned.
+    :param str token: Token provided by ACME server.
+    :param dns.tsigkeyring keyring: TSIG keyring object containg TSIG
+        key pair valid for the provided zone.
+
+    :returns: Add record DNS message.
+    :rtype: dns.message.Message
+
+    """
+    challenge_subdomain = "_acme-challenge.%s" % (zone)
+
+    challenge_request = dns.update.Update(zone, keyring=keyring)
+    # check challenge_subdomain is absent
+    challenge_request.absent(challenge_subdomain)
+    # add challenge_subdomain TXT with token
+    challenge_request.add(
+        challenge_subdomain, constants.DNS_CHALLENGE_TTL, "TXT", token)
+
+    # return req
+    return challenge_request
+
+def del_record(zone, token, keyring): # pylint: disable=unused-argument
+    """Delete record DNS request generator.
+
+    Create a `dns.message.Message` object asking the server to
+    delete a TXT record for the challenge subdomain
+    (`_acme-challenge.example.com`) signed by the provided
+    TSIG keyring.
+
+    :param str zone: Zone (domain) in which the record should exists.
+    :param str token: Not needed
+    :param dns.tsigkeyring keyring: TSIG keyring object containg TSIG
+        keypair valid for the provided zone.
+
+    :returns: Delete record DNS message.
+    :rtype: dns.message.Message
+
+    """
+    challenge_subdomain = "_acme-challenge.%s" % (zone)
```

Comment:
following comment from above, this would read `challenges.DNS.txt_subdomain(zone)`